crates/kubelet/src/config.rs

```diff
@@ -53,6 +54,42 @@ pub struct ServerConfig {
     pub tls_private_key_file: PathBuf,
 }
 
+#[derive(Debug, Default)]
+struct ConfigBuilder {
+    // Some -> Ok(v) = it was present and the value parsed as v
+    //      -> Err(e) = it was present but bad - e described the problem
+    // None = it wasn't present
+    pub node_ip: Option<anyhow::Result<IpAddr>>,
+    pub hostname: Option<String>,
+    pub node_name: Option<String>,
+    pub data_dir: Option<PathBuf>,
+    pub node_labels: Option<HashMap<String, String>>,
+    pub max_pods: Option<anyhow::Result<u16>>,
+    pub server_addr: Option<anyhow::Result<IpAddr>>,
+    pub server_port: Option<anyhow::Result<u16>>,
+    pub server_tls_cert_file: Option<PathBuf>,
+    pub server_tls_private_key_file: Option<PathBuf>,
+}
+
+const NODE_IP_CONFIG_KEY: &str = "node_ip";
+const HOSTNAME_CONFIG_KEY: &str = "hostname";
+const NODE_NAME_CONFIG_KEY: &str = "node_name";
+const DATA_DIR_CONFIG_KEY: &str = "data_dir";
+const NODE_LABELS_CONFIG_KEY: &str = "node_labels";
+const MAX_PODS_CONFIG_KEY: &str = "max_pods";
+const SERVER_ADDR_CONFIG_KEY: &str = "addr";
+const SERVER_PORT_CONFIG_KEY: &str = "port";
+const SERVER_TLS_CERT_FILE_CONFIG_KEY: &str = "tls_cert_file";
+const SERVER_TLS_PRIVATE_KEY_FILE_CONFIG_KEY: &str = "tls_private_key_file";
+
+struct ConfigBuilderFallbacks {
```

Comment:
This struct feels janky...trying to think of a way to do it differently.

Reply:
Let’s give it a try and see what it looks like

Reply:
Right now I feel the completer trait looks messy.  Introducing a completer object increases duplication, and requires either an additional type or an unexpressed invariant.  I'm finding myself getting lost in the code, and I'm not seeing any clarify or consolidation benefit.

I think I have a sense for in what way this struct feels 'janky' and I do sympathise but right now it feels more pragmatic than the completer trait.  Do you want me to continue exploring alternatives, or could we put this aside for now?

Reply:
Let's put it aside for now. Pinging @rylev on this one for any additional thoughts